docs/adr/adr-004-state-interaction.md

```diff
@@ -0,0 +1,123 @@
+(*required for this iteration*) 
+2. **P2P**: discovery of a **bridge** node or other node that is able to provide state (*nice-to-have for this iteration*)
+3. **LOCAL**: eventually, **full** nodes will be able to provide state to the network, and can therefore execute and 
+respond to the queries without relaying queries to celestia-core (*to be scoped out and implemented in later iterations*)
+
+### 1. Core Implementation of `StateAccessor`: `CoreAccess`
+
+`CoreAccess` will be the direct RPC connection implementation of `StateAccessor`. It will use the [lens implementation of ChainClient](https://github.com/strangelove-ventures/lens/blob/main/client/chain_client.go#L23)
+under the hood. `lens.ChainClient` provides a nice wrapper around the Cosmos-SDK APIs. 
+
+Upon node initialisation, the node will receive the endpoint of a trusted celestia-core node. The constructor for 
+`CoreAccess` will construct a new instance of `CoreAccess` with the `lens.ChainClient` pointing at the given 
+celestia-core endpoint.
+
+```go
+type CoreAccess struct {
+    cc *lens.ChainClient
+}
+
+func NewCoreAccess(cc *lens.ChainClient) (*CoreAccess, error) {
+	return &CoreAccess{
+        cc: cc,		
+    }, nil   
+}
+
+// where Account wraps account information and Balance 
+func (ca *CoreAccess) AccountBalance(ctx context.Context, acct Account) (Balance, error) {
+	coins, err := da.cc.QueryBalanceWithAddress(acct.Address)
+	if err != nil {
+		return err
+    }   
+    // returning first index as we only care for account's balance on celestia chain	
+	return coins[0], nil 
+}
+
+// where Tx implements sdk.Msg
+func (ca *CoreAccess) SubmitTx(ctx context.Context, tx Tx) (TxResponse, error) {
+	return da.cc.SendMsg(ctx, tx)
+}
+```
+
+### 2. P2P Implementation of `StateAccessor`: `P2PAccess`
+
+While it is not necessary to detail how `P2PAccess` will be implemented in this ADR, it will still conform to the 
+`StateAccessor` interface, but instead of being provided a core endpoint to connect to via RPC and using `lens.ChainClient`
+for state-related queries, `P2PAccess` will perform service discovery of state-providing nodes in the network and perform
+the state queries via libp2p streams. More details of the p2p implementation will be described in a separate dedicated ADR.
+
+```go
+type P2PAccess struct {
+    // for discovery and communication
+    // with state-providing nodes
+    host host.Host
+    // for managing keys
+    keybase        keyring.Keyring
+    keyringOptions []keyring.Option
+}
+```
+
+### `StateProvider`
+
+A **bridge** node will run a `StateProvider` (server-side of `P2PAccessor`). The `StateProvider` will be responsible for
+relaying the state-related queries through to its trusted celestia-core node.
+
+A new `core.StateFetcher` will be implemented over the `core.Client` in the `core` package. The `core.StateFetcher` will
+contain methods that wrap communication with the `core.Client`. 
```

Comment:
Is it supposed to be used by `StateProvider`? What methods the `core.StateFetcher` will have? Will it wrap lens methods? I see that you want to follow the existing pattern with `core.Fetcher` and I am ok with it, but in this case, it does not seem necessary as the lens is already a wrapper that we can instantiate once in `node/core` and reuse where needed, like in `StateProvider` without additional wrappers. WDYT?

Reply:
So you want StateProvider to use chainclient under the hood to communicate with its RPC client? I think that's totally fine 👍🏻  Pushing changes now.

Reply:
Yes, unless we implement block execution for full nodes.

Reply:
resolved?

Reply:
i assume resolved, sorry need to resolve this to merge!